CDAP-5648 : Dedup Aggregator Plugin

gokulavasan opened the pull request:
JIRA: https://issues.cask.co/browse/CDAP-5648

Note: Will add more tests (failure scenarios) and docs in a followup PR.

gokulavasan committed:
Dedup Aggregator Plugin

gokulavasan committed:
Merge remote-tracking branch 'origin/release/1.3' into feature/dedup-agg-plugin

gokulavasan committed:
Support max, min, first, last functions

Code consolidation

More cleanup and adding config test

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/DedupConfig.java:
```diff
@@ -0,0 +1,101 @@
+  @Nullable
+  private String uniqueFields;
+
+  @Description("Optional field that can be used to choose ")
```
description is incomplete

albertshau replied:
seems like the value here will be something like "name:max(ts)"? If so, this should be changed, the name is not needed at all.  The description should include the syntax we are expecting here.

Also, the property name is misleading.  Based on the name I thought it would just be a field name, but this is more like a record selector or something like that.

albertshau replied:
also, why is this optional?  What happens if it is not specified?

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/DedupConfig.java:
```diff
@@ -0,0 +1,101 @@
+public class DedupConfig extends AggregatorConfig {
+
+  @Description("Optional comma-separated list of fields to perform the distinct on. If none is given, each record " +
+    "will be taken as is. Otherwise, only fields in this list will be considered.")
```
I wrote this description, but its not very descriptive.  Lets change "each record will be taken as is" to "all record fields will be used."

albertshau replied:
This description and property name are not enough for me to understand what this plugin does.  Please update this instead of copy-pasting from distinct.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/DedupConfig.java:
```diff
@@ -0,0 +1,101 @@
+
+    List<FunctionInfo> aggregates = parseAggregation(filterField);
+    if (aggregates.size() != 1) {
+      throw new IllegalArgumentException("Only one filter field is allowed!");
```
nit: we don't use '!' anywhere else, lets replace with a '.'

albertshau replied:
this is also just a confusing error condition. We shouldn't even be trying to parse it as multiple things.  Let's not re-use methods that dont make sense to re-use.  You can refactor to pull out the relevant parsing logic

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/DedupConfig.java:
```diff
@@ -0,0 +1,101 @@
+    public RecordAggregateFunction getAggregateFunction(Schema recordSchema) {
+      Schema.Field recordField = recordSchema.getField(getField());
+      Schema fieldSchema = recordField.getSchema();
+      switch (getFunction()) {
```
function and field can be made protected

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RecordAggregateFunction.java:
```diff
@@ -0,0 +1,34 @@
+ *
+ * @param <T> type of aggregate value
+ */
+public interface RecordAggregateFunction<T> extends AggregateFunction<T> {
```
this is an odd name, its really more like a RecordSelector or a RecordFilter

albertshau replied:
It also shouldn't extend AggregateFunction, because then you require that it does both selection and aggregation which are completely separate things.  If you just need the start, update, finish methods from AggregateFunction, you can create a common interface they both extend.  Things like Max can implement both interfaces.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RecordAggregateFunction.java:
```diff
@@ -0,0 +1,34 @@
+   * @return {@link StructuredRecord} that is chosen based on the aggregate function.
+   */
+  @Nullable
+  StructuredRecord getChosenRecord();
```
If we change this so that it returns a Collection, we can do things like TopK while still allowing min, max, etc.  We also wouldn't have to handle nulls

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/NumberFunction.java:
```diff
@@ -44,7 +45,10 @@ public void beginAggregate() {
           if (val != null) {
-            updateDouble(val.doubleValue());
+            Double updateDouble = updateDouble(val.doubleValue());
+            if (val.doubleValue() == updateDouble) {
```
I'm having trouble understanding this.  Is it to try and fit all AggregateFunctions into RecordAggregateFunctions?  Think we need some refactoring rather than trying to force things that don't fit together.

gokulavasan committed:
Merge remote-tracking branch 'origin/release/1.3' into feature/dedup-agg-plugin

gokulavasan committed:
Addressing PR comments

gokulavasan committed:
Add Max,Min selections

gokulavasan commented:
@albertshau Thanks for the comments. Addressed them. Please take a look again!

gokulavasan committed:
Merge remote-tracking branch 'origin/release/1.3' into feature/dedup-agg-plugin

albertshau commented:
For future reference, it is easier to review if you rebase instead of merge, and squash post-review commits into a single commit.  Then the reviewer can just review that commit.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/DedupAggregator.java:
```diff
@@ -0,0 +1,187 @@
+   * Endpoint request for output schema.
+   */
+  public static class GetSchemaRequest extends DedupConfig {
+    private String inputSchema;
```
learned recently that this can actually be the 'Schema' object so that the plugin doesn't have to parse stringified json.  Should change the other aggregator to Schema as well

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/DedupAggregator.java:
```diff
@@ -0,0 +1,187 @@
+      List<StructuredRecord> outputRecords = selectionFunction.getSelectedRecords();
+      for (StructuredRecord outputRecord : outputRecords) {
+        Schema outputSchema = getOutputSchema(outputRecord.getSchema());
+        validateSchema(outputSchema, uniqueFields, filterFunction);
```
unless I'm misunderstanding, I believe the schema can never be invalid at this point.  In order to get here, all unique fields had to have been present when grouping, and the function had to have worked.

If there is any validation needed, it should happen in the group method.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/DedupConfig.java:
```diff
@@ -0,0 +1,170 @@
+    "that needs to be de-duplicated. This property takes in a field name and the logical operation that needs to be " +
+    "performed on that field on the set of records. The syntax is 'function(field)'. For example, if we want to " +
+    "choose the record with maximum cost for the records with schema 'fname, lname, item, cost', then this field " +
+    "should be set as 'max(cost)'. Note, only one pair of field and function is allowed.")
```
Should add a note about what happens if this is not given.

albertshau replied:
also should mention what functions are supported

gokulavasan commented:
@albertshau Thanks Albert. Will keep that in mind.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RecordFunctionLifecycle.java:
```diff
@@ -0,0 +1,44 @@
+  /**
+   * Called once when the operation is complete.
+   */
+  void finishFunction();
```
From what I saw, it seems like this is always a no-op, can we remove?

gokulavasan replied:
We can but I thought it might be useful in case we make this into a plugin itself. I am ok removing it as well.

albertshau replied:
I think its best to remove until there is a use for it. Easier to introduce a method than take it away.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/NumberSelection.java:
```diff
@@ -0,0 +1,222 @@
+
+  protected abstract void startDouble();
+
+  protected abstract void operateOnInt(StructuredRecord record);
```
would be easier for subclasses if this method got both the int value and the record as arguments, since the first thing they will always do is look up the int value anyway.  Should be able to handle nulls in this class as well so that subclasses don't have to worry about it.

albertshau replied:
similar comment for the other typed methods

albertshau commented:
couple more comments

gokulavasan committed:
Address PR comments

Modified filterOperation to take in a keyvalue format since the necessary UI widget doesn't exist now. Also added widget JSON and docs

gokulavasan commented:
@albertshau Addressed comments. Please review the last commit. 
@awholegunch The last commit in this PR also contains a doc for the new Deduplicate plugin (Deduplicate-batchaggregator.md). Can you please review the same?

albertshau commented:
code lgtm

Commit:
Update Deduplicate-batchaggregator.md

Commit:
Update DedupAggregator.java

Commit:
Update DedupConfig.java

Rewrite to match other docs.

Commit:
Update Deduplicate-batchaggregator.md

Fix line breaks

awholegunch commented:
I've edited the documentation. Please review my changes. Otherwise, LGTM.

Commit:
Update Deduplicate-batchaggregator.md

Make everything optional.

gokulavasan commented:
Thanks @albertshau! Thanks @awholegunch! Doc fixes look good. Will merge after build passes.

gokulavasan committed:
Fixing checkstyle

gokulavasan committed:
Merge remote-tracking branch 'origin/feature/dedup-agg-plugin' into feature/dedup-agg-plugin

gokulavasan commented:
Build passed : http://builds.cask.co/browse/HYP-DBT5-1
